src/stream/decoder.rs

```diff
@@ -142,54 +273,139 @@ impl<R: Read> Read for Decoder<R> {
             pos: self.offset,
         };
 
-        if self.offset > self.buffer.capacity() {
-            // If we've reached the end of the frame before,
-            // don't even try to read more.
-            return Ok(0);
-        }
-
         let mut out_buffer = zstd_sys::ZSTD_outBuffer {
             dst: buf.as_mut_ptr() as *mut c_void,
             size: buf.len(),
             pos: 0,
         };
-        while out_buffer.pos != buf.len() {
 
-            let input_exhausted = in_buffer.pos == in_buffer.size &&
-                                  !self.refill_buffer(&mut in_buffer)?;
-
-            let res = unsafe {
-                let code =
-                    zstd_sys::ZSTD_decompressStream(self.context.s,
-                                              &mut out_buffer as *mut zstd_sys::ZSTD_outBuffer,
-                                              &mut in_buffer as *mut zstd_sys::ZSTD_inBuffer);
-                parse_code(code)?
+        loop {
+            let should_stop = match self.state {
+                DecoderState::Completed => {
+                    return Ok(0);
+                },
+                DecoderState::RefillBuffer(action) => {
+                    self.handle_refill(action, &mut in_buffer, &mut out_buffer)?
+                },
+                DecoderState::Active => {
+                    self.handle_active(&mut in_buffer, &mut out_buffer)?
+                },
             };
+            if should_stop {
+                break;
+            }
+        }
+
+        return Ok(out_buffer.pos);
+    }
+}
 
-            if res > 1 && input_exhausted {
-                // zstd keeps asking for more, but we're short on data!
-                return Err(io::Error::new(io::ErrorKind::UnexpectedEof,
-                                          "incomplete frame"));
+#[cfg(feature = "tokio")]
+impl<R: AsyncRead> AsyncRead for Decoder<R> {
+    unsafe fn prepare_uninitialized_buffer(&self, _buf: &mut [u8]) -> bool {
```

Comment:
Since we never read the output buffer, I suppose it's safe to avoid zeroing it out, improving performance a little bit.